### ISSUE TITLE
EventSource executes deferred commands in the wrong order.  

### ISSUE THREAD

**Issue by vancem** [dotnet/coreclr#14728]
In the attached program that uses EventSource  [Program.cs.txt](https://github.com/dotnet/coreclr/files/1423398/Program.cs.txt) should emit the message 'Hey!'  but does not.  

The reason is that EventSource has the concept of a 'deferred' command and these commands are executed in the wrong order (they are executed last first).    Basically if an EventListener exists before an EventSource exists, then when that EventSource gets created, it calls the EventListener to ask if wants to subscribe.    However because the EventSource is being created, it is not actually ready for the commands the EventListener might give to it.    So it remembers these commands and executes them just after it is fully initialized.   

There is bug in this logic, as the commands are inserted on linked list that naturally forms a STACK (see  https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Diagnostics/Tracing/EventSource.cs#L2682).  The result is that the deferred commands are executed in reverse order.  

It is a easy fix (insert the new entries at the end of the list rather than the beginning).   

Work-arounds:  Not that different EventListeners have their own set of filtering state, so the order AMONG eventListeners is not important (there are no guarentees there).   Thus this only becomes an issue if you execute multiple commands in the OnEventSourceCreated (which is unusual).  Also this is only a problem for deferred commands.  If the EventSource exists before the EventListener, there is also no problem (thus in the original repro, if you put a 'AlphaSource.Log.IsEnabled() in the begnning of main (which forces the EventSource to be created), the problem will go away. 

This is reported to work properly on Desktop, but it is unclear why as the code is the same (I think).   It should be fixed in both runtimes.

**Comment by vancem** [comment]
We should also add regression test for this.

**Comment by brianrob** [comment]
@vancem, I just tried this on desktop and it does not repro.  This implies that the EventSource is fully initialized (or thinks that it is).  I will debug this to see if I can figure out where the behavior difference is.  In the meantime, I think the change in #14729 is a good thing to be doing anyway.

**Comment by vancem** [comment]
I had a hunch on why it does not repro on Desktop.   The issue is that Desktop is more aggressive about running class constructors early (the EventSource needs to be created after the EventListener to repro).  Changing the program so that we don't initialize the 'Log' in a class constructor but in the body of main (as shown below). 
```
    static void Main(string[] args)
    {
        using (var al = new AlphaListener())
        {
            Console.WriteLine("Creating the AlphaSource");
            AlphaSource.Log = new AlphaSource();   // Create the source AFTER the Listener

            Console.WriteLine("given the events we enabled only the 'Message' event should show up");
            // However we don't see it (because the EnableEvents are executed in reverse order).
            AlphaSource.Log.Tricky(1, "Foo", "Bar");     // Not captured because it is only Level=Informational
            AlphaSource.Log.Message("Hey!");             // Captured
            AlphaSource.Log.DebugMessage("Yo!");         // Not captured because Level=Verbose


            Console.WriteLine("Waiting for a key to be hit");
            Console.ReadKey();
        }
    }
```
causes it to repro on .NET Desktop.    

This kind of things is less likely to happen (we tell people to use statics for EventSources).   Still I think it is a safe change and is worth back-porting if you can do it along with other changes.

**Comment by brianrob** [comment]
Fixed via #14729.
